File: NeetCode150/Binary Trees/Kth Smallest Element in a BST (230).py
# Iterative Solution
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        s = []
        cur = root
        while s or cur:
            if cur:
                s.append(cur)
                cur = cur.left
            else:
                cur = s.pop()
                k -= 1
                if not k:
                    return cur.val
                cur = cur.right
       
        return cur.val

File: NeetCode150/Binary Trees/Kth Smallest Element in a BST (230).cpp
#include <bits/stdc++.h>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
    public:
        int kthSmallest(TreeNode* root, int& k) {
            if(!root)   return -1;
            int l = kthSmallest(root->left, k);
            if(l != -1)     return l;
            k--;
            if(!k)  return root->val;
            int r = kthSmallest(root->right, k);
            if(r != -1)     return r;
            return -1;
        }
    
};